junit5/src/main/java/cz/xtf/junit5/listeners/OpenshiftStateRecorder.java

```diff
@@ -0,0 +1,199 @@
+ *    To use this interface add this class as a listener
+ *
+ * LifecycleMethodExecutionExceptionHandler
+ *    This interface is executed when @BeforeAll or @BeforeEach method fails - throw an exception.
+ *    To use this interface a test class needs to be extended with this class using @ExtendWith
+ */
+@Slf4j
+public class OpenshiftStateRecorder implements TestExecutionListener, LifecycleMethodExecutionExceptionHandler {
+	private static final Path statusDir = Paths.get("log","status");
+
+	private final OpenShift openShift = OpenShifts.master();
+	private final OpenShift buildOpenShift;
+
+	private final String namespace;
+	private final String buildNamespace;
+
+	public OpenshiftStateRecorder(){
+		namespace = openShift.getNamespace();
+		buildNamespace = XTFConfig.get(BuildManagerConfig.BUILD_NAMESPACE);
+		buildOpenShift = OpenShifts.master(buildNamespace);
+	}
+
+	private String getTestDisplayName(TestIdentifier testIdentifier) {
+		String className = testIdentifier.getParentId().get()
+			.replaceAll(".*class:", "")
+			.replaceAll("].*", "");
+		return String.format("%s#%s", className, testIdentifier.getDisplayName());
+	}
+
+	// Log the openshift state if BeforeAll fail
+	@Override
+	public void handleBeforeAllMethodExecutionException(final ExtensionContext context, final Throwable ex) throws Throwable {
+		try {
+			storeOpenshiftState(context.getDisplayName());
+		} catch(Exception newEx){
+			// if new exception is thrown, add it to original one so it is not lost
+			ex.addSuppressed(newEx);
+		}
+		throw ex;
+	}
+
+	// Log the openshift state if BeforeEach fail
+	@Override
+	public void handleBeforeEachMethodExecutionException(final ExtensionContext context, final Throwable ex) throws Throwable {
+		try {
+			storeOpenshiftState(context.getDisplayName());
+		} catch(Exception newEx){
+			// if new exception is thrown, add it to original one so it is not lost
+			ex.addSuppressed(newEx);
+		}
+		throw ex;
+	}
+
+	// Log the openshift state if test fail
+	@Override
+	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
+		try {
+			if (FAILED.equals(testExecutionResult.getStatus()) && testIdentifier.isTest()) {
+				storeOpenshiftState(getTestDisplayName(testIdentifier));
+			}
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+
+	private void storeOpenshiftState(String testIdentifier) throws IOException {
+		final Path testLogDir = statusDir.resolve(testIdentifier);
+		Files.createDirectories(testLogDir);
+
+		storeNamespaceState(testLogDir, openShift);
+		storePods(testLogDir, openShift);
+
+		if (!namespace.equals(buildNamespace)){
+			storeNamespaceState(testLogDir, buildOpenShift);
+			storePods(testLogDir, buildOpenShift);
+		}
+	}
+
+	private void storePods(Path dir, OpenShift openshift){
+		for (Pod pod : openshift.getPods()) {
+			try {
+				openshift.storePodLog(pod, dir, pod.getMetadata().getName() + ".log");
```

Comment:
In xtf there is already `PodLogRecorder` to store logs once test failed. 

What does it adds on top of that?

Reply:
Nothing from pod logs themselfs. Why I added it to this is:
- It also records from build namespace.
- It is also recorded when Before* method fails. So it's possible to get logs from builds, non-tested services and other thing that can fail in setup.